### PR TITLE
build: propagate ASAN/TSAN flags down to cmake_external deps.

### PR DESCRIPTION
**Pull request by htuch**
While we await a solution for
https://github.com/bazelbuild/rules_foreign_cc/issues/154#issuecomment-466504751, this PR provides a
temporary workaround to get full ASAN/TSAN propagation as needed.

Risk level: Low
Testing: --config=clang-{tsan,asan}. This was broken strangely this morning on my distribution,
after some clang-7 package upgrade this morning, due to the mismatched linker/compiler flags we
previously had been sending to cmake_external builds.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
CC @mergeconflict @irengrig

**Comment by lizan** [comment]
@htuch seems CI is failing, maybe you need newer image?

**Comment by htuch** [comment]
@lizan seems some mystery compile error in `gperftools`, no idea why it's suddenly missing `munmap`, will dig.

**Comment by htuch** [comment]
Hmm, `checking for working mmap... no` from the configure log, that's probably not good.

**Comment by PiotrSikora** [comment]
> Hmm, `checking for working mmap... no` from the configure log, that's probably not good.

You might need to update autoconf:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=866406

**Comment by htuch** [comment]
@PiotrSikora thanks for that pointer, I think we have the same issue.

From config.log:
```
configure:18376: clang -o conftest -fsanitize=address,undefined -fno-sanitize=vptr -ggdb3 -fno-omit-frame-pointer -O2 -DNDEBUG  -lpthread conftest.c  >&5
configure:18376: $? = 0
configure:18376: ./conftest

=================================================================
==2690==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4096 byte(s) in 1 object(s) allocated from:
    #0 0x4c2b43 in __interceptor_malloc (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_deps_cache_b22e04bff96538ea37e715942da6315c/gperftools.dep.build/gperftools-fc00474ddc21fff618fc3f009b46590e241e425e/conftest+0x4c2b43)
    #1 0x4f40a8 in main /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_deps_cache_b22e04bff96538ea37e715942da6315c/gperftools.dep.build/gperftools-fc00474ddc21fff618fc3f009b46590e241e425e/conftest.c:217:20
    #2 0x7f8d3fa5882f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291

Direct leak of 4096 byte(s) in 1 object(s) allocated from:
    #0 0x4c2b43 in __interceptor_malloc (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_deps_cache_b22e04bff96538ea37e715942da6315c/gperftools.dep.build/gperftools-fc00474ddc21fff618fc3f009b46590e241e425e/conftest+0x4c2b43)
    #1 0x4f3c72 in main /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_deps_cache_b22e04bff96538ea37e715942da6315c/gperftools.dep.build/gperftools-fc00474ddc21fff618fc3f009b46590e241e425e/conftest.c:168:19
    #2 0x7f8d3fa5882f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291

SUMMARY: AddressSanitizer: 8192 byte(s) leaked in 2 allocation(s).
```

.. and we seem to have an autoconf version prior to the bug fix ...
```
++ 1551140675.668187254 apt-cache show autoconf
Package: autoconf
Status: install ok installed
Priority: optional
Section: devel
Installed-Size: 1861
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Architecture: all
Multi-Arch: foreign
Version: 2.69-9
...
```

Will look into updating the base image.

File: bazel/repositories.bzl [hunk 1]
```diff
@@ -73,6 +73,10 @@ def _build_recipe_repository_impl(ctxt):
         )
     ctxt.symlink(Label("//ci/prebuilt:BUILD"), "BUILD")
 
+    # Data files for build recipes
+    for f in ["gperftools.patch.gz"]:
```
Comment (PiotrSikora) [inline bazel/repositories.bzl]: Wouldn't it be easier to update autoconf, or simply hardcode `-DHAVE_MMAP` to ignore the breaking test?

Comment (htuch) [reply bazel/repositories.bzl]: @lizan expressed a preference for this approach; for release versions we have fixed autoconf outputs, we will revert back to this when we next move to a release version.

**Comment by htuch** [comment]
Looks like a legit TSAN failure, libevent does some silly unsafe things by writing to a shared boolean across threads (`event_debug_mode_too_late`). Filed https://github.com/libevent/libevent/issues/777, will disable this debug support for TSAN builds.

**Comment by htuch** [comment]
I have a workaround for the libevent TSAN issues, next up is the ASAN issue in LuaJIT at https://circleci.com/gh/envoyproxy/envoy/172877?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link.

**Comment by htuch** [comment]
More TSAN errors, this time in dispatcher posting, will dig further tomorrow.

File: .bazelrc [hunk 2]
```diff
@@ -13,6 +13,12 @@ build:asan --define ENVOY_CONFIG_ASAN=1
 build:asan --copt -fsanitize=address,undefined
 build:asan --linkopt -fsanitize=address,undefined
 build:asan --copt -fno-sanitize=vptr
+# TODO(htuch): these are temporary workarounds, remove when
```
Comment (lizan) [inline .bazelrc]: nit: It's not clear which are "these", put them in the bottom?

**Comment by htuch** [comment]
Uh oh, with the bump to Clang 7.0.1 in https://github.com/envoyproxy/envoy/pull/5953 (and basically anyone else who wants to bump the CI image), we will hit failures like https://github.com/envoyproxy/envoy/pull/5953.

@lizan is LuaJIT fixed with your LSAN_OPTIONS change? I think maybe we only have a some libevent TSAN issues re: atomics.

**Comment by htuch** [comment]
Closing out in favor of cleaned up https://github.com/envoyproxy/envoy/pull/6196.

**Pull request closed without merging**